File: 백준/Gold/2252. 줄 세우기/줄 세우기.py
import sys
input = sys.stdin.readline
from collections import deque

n, m = map(int, input().split())
maps = [[] for _ in range(n+1)]
dist = [0 for i in range(n+1)]
answer = []
for i in range(m):
    a, b = map(int, input().split())
    maps[a].append(b)
    dist[b] += 1

def topologySort():
    que = deque([])
    for i in range(1, n+1):
        if not dist[i]:
            que.append(i)

    for i in range(n):
        if not que:
            return
        node = que.popleft()
        for nextnode in maps[node]:
            dist[nextnode] -= 1
            if not dist[nextnode]:
                que.append(nextnode)

        answer.append(node)


topologySort()

print(*answer)

File: 백준/Gold/2252. 줄 세우기/줄 세우기.cc
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

int n, m;
vector <vector<int>> graph(100001);
int dist[100001];

void topologySort() {
    queue<int> que;
    for(int i = 1; i <= n; i++) {
        if(!dist[i])que.push(i);
    }
    for(int i = 0; i < n; i++) {
        if(que.empty()){
            return;
        }
        int node = que.front();
        que.pop();
        for(int i = 0; i< graph[node].size(); i++) {
            int nextNode = graph[node][i];
            dist[nextNode]--;
            if(!dist[nextNode])que.push(nextNode);
        }
        cout<<node<<" ";
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n >> m;
    for(int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        dist[b]++;
    }
    topologySort();

    return 0;
}